### ISSUE TITLE
for loop dropping data

### ISSUE THREAD

**Issue by mike-bales** [JHU-DataViz-Summer16/final-projects#1]
I'm trying to put all three of my charts into a single SVG so I can do transformations between them. I'm using a for loop to filter the data and create each chart. I've basically [got the loop working](https://github.com/mike-bales/police-fatality-data-viz/blob/for-loop-broken/script.js#L52) except for one issue. On the second and third passes through the loop the `[0]` element is being dropped in my data. This element corresponds to the December 31 `<g>` element. Using `console.log()` you can see that is present in the `vicsByDate` variable but then missing in the `dateGroup` variable. Any help figuring out why this happening would be appreciated. 

You can use the link above to get to the offending loop or you can go to the script.js file in the for-loop-broken branch of my final project repo.

**Comment by cmgiven** [comment]
On line 86, you set the key to use the property `deathDate`. Try console.log()ing that in the accessor function. Can you see why that drops the first element of future enter groups?

By the way, this could point to an advantage of appending each year's points in a separate group. Instead of selecting all days every time you run through the loop, you could select the year group, and then just operate on the days within that year.

**Comment by mike-bales** [comment]
I'm not sure what you mean by 

> Try console.log()ing that in the accessor function.

I'm still unclear on terms like accessor function and what they refer to.

Do you mean like this?

``` javascript
    var dateGroup = chart.svg.selectAll(".day")
                             .data(vicsByDate, function (d) { console.log(d.deathDate); return d.deathDate;})
                             .enter().append("g")
                             .attr("transform", "translate(0," + (((height+margin.top) * (i+2)) - cellSize) + ")")
                             .attr("class", "day");
```

**Comment by cmgiven** [comment]
Yup, that's exactly what I mean. Sorry for getting jargon-y. By accessor function I mean the function that accesses a property of each datum, or row of data.

**Comment by mike-bales** [comment]
Please keep the jargon coming! I want to learn this stuff so even if I can't exactly call myself a web developer yet I can at least communicate with them intelligently. Unless of course you get sick of me asking you to explain terms. ;)

So that definitely helped me figure out that I was using the wrong property name since it just returned undefined. I should have been using `d.key` since that is what comes out of `.nest()` as the key for each group, not the original property name I used to group the variables. 

But I'll be honest I'm still a little confused on why I was getting the exact behavior I was getting. The function for the key was returning `undefined` but it didn't seem to be matching that missing element to anything since the data for the existing elements weren't changing. I figured that that missing element needed to be bound to either the update, enter, or exit groups but I can't figure out where it went.

**Comment by cmgiven** [comment]
So remember that .enter() creates a selection of elements that don't already exist. This get's complicated, though, when you have duplicate keys.

Let's say you have data A, B, C.
Then you bind data A, B, C, D.
The resulting enter selection would be D, the only element that didn't previously exist.

But you had data undefined, undefined, undefined.
And then you bound data undefined, undefined, undefined, undefined.

D3 took a look at that first undefined and went, "Hey, I already have this! No need to select it."
But then it saw that the remainder of the data used a duplicate key, so it didn't even bother to check whether it had it, and just added it to the enter selection.

Does that help?
